Add MD-SAPT to MDAKits

ALescoulie opened the pull request:
My MDAKit [MD-SAPT](https://github.com/calpolyccg/MDSAPT) runs SAPT calculations over MD trajectories using Psi4. We currently are working on publishing it in JOSS and it is ready to be submitted as an official MDA kit.

ALescoulie committed:
Create metadata.yaml

IAlibay commented:
@ALescoulie we're in the process of improving our docs for adding mdakits, especially when it comes to adding things so that tests will run properly. We'll come back to this (and other PRs open) once that's done, so that we can point folks to the right steps to getting everything working.

orbeckst commented:
FYI, we have the MDAKits tutorial with the section [Submitting a MDAKit to the registry](https://mdakits.mdanalysis.org/makingakit.html#submitting-the-kit-to-the-registry).

orbeckst commented on mdakits/mdsapt/metadata.yaml:
```diff
@@ -0,0 +1,59 @@
+mdanalysis_requires: ">=2.0.0"
+## List(str): a list of commands to use when attempting to run the MDAKit's tests
+run_tests:
+  - pytest --pyargs propkatraj.tests
```
change!

orbeckst commented on mdakits/mdsapt/metadata.yaml:
```diff
@@ -0,0 +1,59 @@
+## List(str): a list of commands to use to install the necessary dependencies required
+## to run the MDAKit's tests
+test_dependencies:
+project_org: Becksteinlab 
```
change!

orbeckst commented on mdakits/mdsapt/metadata.yaml:
```diff
@@ -0,0 +1,59 @@
+## str: the development status of the MDAKit
+development_status: Mature
+## List(str) a list of publications to cite when using the MDAKit
+publications:
```
paper published?

ALescoulie replied:
not yet, were working on it

orbeckst commented on mdakits/mdsapt/metadata.yaml:
```diff
@@ -0,0 +1,59 @@
+development_status: Mature
+## List(str) a list of publications to cite when using the MDAKit
+publications:
+community_home:
```
Maybe enable GitHub discussions on https://github.com/calpolyccg/MDSAPT ??

ALescoulie replied:
I enabled discussions

orbeckst replied:
Can you add an link to discussions?

orbeckst commented on mdakits/mdsapt/metadata.yaml:
```diff
@@ -0,0 +1,59 @@
+  - protein
+  - quantum
+## str: the license the mdakit falls under
+license: GPL-2.0-or-later
```
According to your repo, it's GPL 3.0 or later

orbeckst reviewed (changes requested):
Meta data requires some obvious changes, in particular the ones to make tests run.

orbeckst commented:
@ALescoulie could you address the changes above? They are really minor!

orbeckst committed:
Update metadata.yaml

orbeckst commented on mdakits/mdsapt/metadata.yaml:
```diff
@@ -0,0 +1,59 @@
+## release of the code. Note: only one installation method can currently
+## be defined. We suggest using conda/mamba where possible.
+install:
+  - mamba install -c psi4 mdsapt
```
according to https://psicode.org/psi4manual/master/conda.html, psi4 is now on conda-forge

orbeckst replied:
However, mdsapt is only in `-c psi4`.

orbeckst commented on mdakits/mdsapt/metadata.yaml:
```diff
@@ -0,0 +1,59 @@
+## List(str): a list of commands to use when installing the mdakit from its
+## source code.
+src_install:
+  - pip install git+https://github.com/calpolyccg/MDSAPT@master
```
Install pre-requisites
```
  # required dependencies from environment.yaml
  - mamba install -c psi4 -c conda-forge -c defaults "psi4>=1.6.1,<1.7" click numpy openmm pandas pdbfixer pytest pydantic pytest-cov pyyaml rdkit
  - pip install git+https://github.com/calpolyccg/MDSAPT@master
```

orbeckst committed:
install mdsapt properly

- conda package requires dev label
- install pre-requisites for source installation

orbeckst commented:
The tests don't like to be run with `--pyargs`
```
Run tests=$(python utils/get_runtests.py --mdakit mdsapt --runtype latest)
tests: pytest --pyargs mdsapt.tests
ERROR: module or package not found: mdsapt.tests (missing __init__.py?)
```
Not sure if `mdsapt` is not installed properly or if it's something else; the tests directory contains `__init__.py` https://github.com/calpolyccg/MDSAPT/tree/master/mdsapt/tests.

orbeckst committed:
get tests from checked out source

orbeckst commented:
The tests fail because some issues with pydantic. This looks as if it needs to be fixed in MD-SAPT
```
==================================== ERRORS ====================================
_________________ ERROR collecting mdsapt/tests/test_config.py _________________
/usr/share/miniconda3/envs/test/lib/python3.9/site-packages/_pytest/runner.py:341: in from_call
    result: Optional[TResult] = func()
/usr/share/miniconda3/envs/test/lib/python3.9/site-packages/_pytest/runner.py:372: in <lambda>
    call = CallInfo.from_call(lambda: list(collector.collect()), "collect")
/usr/share/miniconda3/envs/test/lib/python3.9/site-packages/_pytest/python.py:531: in collect
    self._inject_setup_module_fixture()
/usr/share/miniconda3/envs/test/lib/python3.9/site-packages/_pytest/python.py:545: in _inject_setup_module_fixture
    self.obj, ("setUpModule", "setup_module")
/usr/share/miniconda3/envs/test/lib/python3.9/site-packages/_pytest/python.py:310: in obj
    self._obj = obj = self._getobj()
/usr/share/miniconda3/envs/test/lib/python3.9/site-packages/_pytest/python.py:528: in _getobj
    return self._importtestmodule()
/usr/share/miniconda3/envs/test/lib/python3.9/site-packages/_pytest/python.py:617: in _importtestmodule
    mod = import_path(self.path, mode=importmode, root=self.config.rootpath)
/usr/share/miniconda3/envs/test/lib/python3.9/site-packages/_pytest/pathlib.py:567: in import_path
    importlib.import_module(module_name)
/usr/share/miniconda3/envs/test/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1030: in _gcd_import
    ???
<frozen importlib._bootstrap>:1007: in _find_and_load
    ???
<frozen importlib._bootstrap>:972: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:228: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1030: in _gcd_import
    ???
<frozen importlib._bootstrap>:1007: in _find_and_load
    ???
<frozen importlib._bootstrap>:972: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:228: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1030: in _gcd_import
    ???
<frozen importlib._bootstrap>:1007: in _find_and_load
    ???
<frozen importlib._bootstrap>:986: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:680: in _load_unlocked
    ???
<frozen importlib._bootstrap_external>:850: in exec_module
    ???
<frozen importlib._bootstrap>:228: in _call_with_frames_removed
    ???
mdsapt/__init__.py:8: in <module>
    from .config import Config, load_from_yaml_file
mdsapt/config.py:132: in <module>
    class RangeFrameSelection(BaseModel):
mdsapt/config.py:145: in RangeFrameSelection
    @root_validator()
/usr/share/miniconda3/envs/test/lib/python3.9/site-packages/pydantic/deprecated/class_validators.py:228: in root_validator
    raise PydanticUserError(
E   pydantic.errors.PydanticUserError: If you use `@root_validator` with pre=False (the default) you MUST specify `skip_on_failure=True`. Note that `@root_validator` is deprecated and should be replaced with `@model_validator`.
E   
E   For further information visit https://errors.pydantic.dev/2.3/u/root-validator-pre-skip
```

orbeckst reviewed (changes requested):
There are problems with pydantic that makes the tests fail, at least with the versions that are used in the testing here. I raised https://github.com/calpolyccg/MDSAPT/issues/42 . When the testing code is changed, we can re-run here. 

Please ping me.

orbeckst commented on mdakits/mdsapt/metadata.yaml:
```diff
@@ -0,0 +1,62 @@
+## source code.
+src_install:
+  # required dependencies from environment.yaml
+  - mamba install -c psi4 -c conda-forge -c defaults "psi4>=1.6.1,<1.7" click numpy openmm pandas pdbfixer pytest pydantic pytest-cov pyyaml rdkit
```
Question for @IAlibay : Is it correct to do the installation of dependencies _before_ the source pip install if I don't want dependencies to be pip-installed? I'd love to be able to say `mamba install -f environment.yml` but to my knowledge that doesn't work (only works with `mamba create`). Because I have to list dependencies explicitly, I also have to explicitly add pinning (like for `psi4`). That looks a bit awkward and prone to breakage. Is there a better way?

IAlibay replied:
🤔 yes that is indeed a bit of a problem.

What would the desired workflow be here? Fetch the yaml file from the remote directory and create the environment using that? Or should we try to read the yaml file and install from there?

This is definitely something to be considered as part of the refactor.

orbeckst replied:
At the end of the day the package developers should be responsible for keeping the meta data in sync with the package installation. I’d rather keep our side simple and suggest that ultimately packages follow standard procedures. In the meantime explicit pinning should be fine as a temporary measure — if this gets the kit installed correctly.

ALescoulie committed:
Merge branch 'MDAnalysis:main' into main

ALescoulie committed:
add conda-forge install

ALescoulie commented:
I pushed an update that has the conda forge install, I'd appriacte if someone could take a look at this.

IAlibay commented on mdakits/mdsapt/metadata.yaml:
```diff
@@ -0,0 +1,63 @@
+## source code.
+src_install:
+  # required dependencies from environment.yaml
+  mamba install -c conda-forge psi4 >= 1.9.1, mdanalysis >=2.7.0, rdkit >=2023.09.5, openmm >=8.1.1, pdbfixer >=1.9, numpy, click, pandas, pyarrow, pyyaml
+  pip install git+https://github.com/calpolyccg/MDSAPT@master
```
```suggestion
  - mamba install -c conda-forge psi4 >= 1.9.1, mdanalysis >=2.7.0, rdkit >=2023.09.5, openmm >=8.1.1, pdbfixer >=1.9, numpy, click, pandas, pyarrow, pyyaml
  - pip install git+https://github.com/calpolyccg/MDSAPT@master
```

IAlibay reviewed (commented):
Sorry just giving this a very brief look, the `src_install` field expects a list, added a suggested change to fix that.

ALescoulie committed:
make src-install into a list

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

ALescoulie committed:
fix test dependencies

ALescoulie commented:
Icommitted @IAlibay's change to dependency install command, and fixed an issue with the test dependencies command.

@IAlibay  and @orbeckst is there anything else to fix for this PR?

ALescoulie committed:
fix comma typo in test deps install

ALescoulie commented:
The CI failed because I put a comma between test dependencies on the `mamba install` line which caused it to fail. I pushed a fix, @orbeckst could you try running the full CI again?

orbeckst commented:
There was a fail for develop and I restarted it. Not sure if this is what you meant.

orbeckst reviewed (changes requested):
The develop CI is still failing https://github.com/MDAnalysis/MDAKits/actions/runs/9766164949/job/26961722361?pr=47

Does the command
```
mamba install -c conda-forge psi4 >= 1.9.1, mdanalysis >=2.7.0, rdkit >=2023.09.5, openmm >=8.1.1, pdbfixer >=1.9, numpy, click, pandas, pyarrow, pyyaml;pip install git+https://github.com/calpolyccg/MDSAPT@master
```
work when you try it manually?

orbeckst commented:
(Is it possible that there should be quotation marks in the install command?)

IAlibay commented on mdakits/mdsapt/metadata.yaml:
```diff
@@ -0,0 +1,64 @@
+## source code.
+src_install:
+  # required dependencies from environment.yaml
+  - mamba install -c conda-forge psi4 >= 1.9.1, mdanalysis >=2.7.0, rdkit >=2023.09.5, openmm >=8.1.1, pdbfixer >=1.9, numpy, click, pandas, pyarrow, pyyaml
```
```suggestion
  - mamba install -c conda-forge "psi4>= 1.9.1" "mdanalysis >=2.7.0" "rdkit >=2023.09.5" "openmm >=8.1.1" "pdbfixer >=1.9" numpy click pandas pyarrow pyyaml
```

orbeckst replied:
let's try it, thx

IAlibay reviewed (commented):
I think this should work

orbeckst committed:
Update mdakits/mdsapt/metadata.yaml

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

orbeckst reviewed (changes requested):
Develop installation works but you get a segfault
```
Thread 0x00007f8ad06fe640/home/runner/work/_temp/e522ce51-b634-481c-b233-3008ac944e02.sh: line 3:  3289 Segmentation fault      (core dumped) pytest -v ./mdsapt/tests
mdsapt/tests/test_sapt.py::TestSAPT::test_run_traj_sapt 
```
Can you look into what's happening there? Is this locally reproducible?

ALescoulie commented:
Yeah I'll look at that later day. The issue probably is that need to bring develop up to date with master because we've pushed a lot of the recent fixes straight from feature branch into master.

orbeckst commented on mdakits/mdsapt/metadata.yaml:
```diff
@@ -0,0 +1,64 @@
+src_install:
+  # required dependencies from environment.yaml
+  - mamba install -c conda-forge "psi4>= 1.9.1" "mdanalysis >=2.7.0" "rdkit >=2023.09.5" "openmm >=8.1.1" "pdbfixer >=1.9" numpy click pandas pyarrow pyyaml
+  - pip install git+https://github.com/calpolyccg/MDSAPT@master
```
Note that the „develop“ CI installs your source (ie your master branch) and tests against the MDAnalysis develop branch! That’s the configuration to check for the segfault.

orbeckst reviewed (changes requested):
Fix segfault in develop CI. 

(Excuse brevity, please)

orbeckst commented:
Have you made any progress in figuring out the reason for the segfault? Is it even locally reproducible?

ALescoulie commented:
@orbeckst I haven't gotten around to trying yet, I forgot to mention on this thread that I just moved (I'm still in the bay area, my partner and I found a great deal on a place in SF, so jumped on it) and have been busy with that. Now that I finally set up my desktop again, I'll try and replicate it in a container since I use NixOS and don't want to potentially introduce Nix related problems to the process.

orbeckst commented:
Ping @ALescoulie ;-) — did you have some opportunities to dig into MD-SAPT again?

orbeckst commented:
@ALescoulie are you still looking into registering MD-SAPT or should we close the PR for now?

orbeckst commented:
@ALescoulie given that the PR has been sitting here for a while, I am going to close it. If you want to come back to please say something and I am happy to reopen it.
